datajunction-server/datajunction_server/api/helpers.py

```diff
@@ -821,3 +827,65 @@ def build_sql_for_multiple_metrics(  # pylint: disable=too-many-arguments,too-ma
         engine,
         leading_metric_node.current.catalog,
     )
+
+
+async def query_event_stream(  # pylint: disable=too-many-arguments
+    query: QueryWithResults,
+    query_service_client: QueryServiceClient,
+    columns: List[Column],
+    request,
+    timeout: float = 0.0,
+    stream_delay: float = 0.5,
+    retry_timeout: int = 5000,
+):
+    """
+    A generator of events from a query submitted to the query service
+    """
+    starting_time = time.time()
+    # Start with query and query_next as the initial state of the query
+    query_prev = query_next = query
+    query_id = query_prev.id
+    _logger.info("sending initial event to the client for query %s", query_id)
+    yield {
+        "event": "message",
+        "id": uuid.uuid4(),
+        "retry": retry_timeout,
+        "data": json.dumps(query.json()),
+    }
+    # Continuously check the query until it's complete
+    while not timeout or (time.time() - starting_time < timeout):
+        # Check if the client closed the connection
+        if await request.is_disconnected():  # pragma: no cover
+            _logger.error("connection closed by the client")
+            break
+
+        # Check the current state of the query
+        query_next = query_service_client.get_query(query_id=query_id)  # type: ignore
+        if query_next.state in END_JOB_STATES:
+            _logger.info(
+                "query end state detected (%s), sending final event to the client",
+                query_next.state,
+            )
+            if query_next.results.__root__:  # pragma: no cover
+                query_next.results.__root__[0].columns = columns or []
+            yield {
+                "event": "message",
+                "id": uuid.uuid4(),
+                "retry": retry_timeout,
+                "data": json.dumps(query_next.json()),
+            }
+            _logger.info("connection closed by the server")
+            break
+        if query_prev != query_next:  # pragma: no cover
```

Comment:
What's an example of a case where it reaches this part? Is this for handling when the query's state changes but it hasn't reached one of the end states?

Reply:
Yeah that's exactly right. The server constantly checks on the query state on an interval defined by `stream_delay` (defaults to half a second). But this prevents the server from sending an event to the client when the data is exactly the same as the last event. The usual flow of a successful query would then be 3 events of `ACCEPTED`, `SCHEDULED`, and `FINISHED`.